### PR TITLE
executor: reduce memory allocation in the initialization of hash agg executor

### LINKED ISSUE #53898: (*HashAggExec).initForParallelExec is too slow
## Enhancement

(*HashAggExec).initForParallelExec is too slow. and it take many time to alloc the memory for ```Chunk```.

https://flamegraph.com/share/7c32d2ec-2648-11ef-aba3-6a3d1814cbe4

image

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #53898

Problem Summary:

### What changed and how does it work?

image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/aggregate/agg_hash_executor.go [hunk 1]
```diff
@@ -304,12 +304,13 @@ func (e *HashAggExec) initPartialWorkers(partialConcurrency int, finalConcurrenc
 			globalOutputCh:       e.finalOutputCh,
 			partialResultsMap:    partialResultsMap,
 			groupByItems:         e.GroupByItems,
-			chk:                  exec.TryNewCacheChunk(e.Children(0)),
-			groupKeyBuf:          *groupKeyBuf,
-			serializeHelpers:     aggfuncs.NewSerializeHelper(),
-			isSpillPrepared:      false,
-			spillHelper:          e.spillHelper,
-			inflightChunkSync:    e.inflightChunkSync,
+			// chk:                  e.NewChunkWithCapacity(e.RetFieldTypes(), 0, e.MaxChunkSize()),
```
Comment (hawkingrei) [inline pkg/executor/aggregate/agg_hash_executor.go:307]: Can you remove it ?

Comment (xzhangxian1008) [reply pkg/executor/aggregate/agg_hash_executor.go:307]: > Can you remove it ?

The ci always fails, I'm trying to figure out which leads to the fail.

**Review by hawkingrei — APPROVED** [review]

File: pkg/util/chunk/chunk.go [hunk 2]
```diff
@@ -187,12 +187,24 @@ func (c *Chunk) MemoryUsage() (sum int64) {
...
-		sum += curColMemUsage
+		sum += int64(unsafe.Sizeof(*col)) + int64(cap(col.nullBitmap)) + int64(cap(col.offsets)*8) + int64(cap(col.data)) + int64(cap(col.elemBuf))
 	}
 	return
 }
 
+// GetMemoryUsageCap returns the total memory usage that may be used in the future.
+// We ignore the size of Column.length and Column.nullCount
+// since they have little effect of the total memory usage.
+func (c *Chunk) GetMemoryUsageCap(capacity int) (sum int64) {
```
Comment (XuHuaiyu) [inline pkg/util/chunk/chunk.go:198]: Why not use chunk.MemoryUsage, what's the difference?

**Review by XuHuaiyu — COMMENTED** [review]
Any test result after this improvement can be shown?

File: pkg/util/chunk/chunk.go (hunk 2 above)
Comment (xzhangxian1008) [reply pkg/util/chunk/chunk.go:198]: > Why not use chunk.MemoryUsage, what's the difference?

`chunk.MemoryUsage` returns the memory that has been allocated. However, we don't allocate any memory when the chunk is initialized. So, we need a new function to return the memory that the chunk may be allocated in the future.

**Comment by xzhangxian1008** [comment]
> Any test result after this improvement can be shown?

test result has been updated in the pr description.

**Review by windtalker — APPROVED** [review]
lgtm

**Pull request merged**